feat(android): Add custom masking options for session replay (#11624)
getsentry/sentry-docs: docs/platforms/android/session-replay/privacy/index.mdx

Before publising an App with Session Replay enabled, make sure to test it thoroughly to ensure that no sensitive data is exposed.

If you choose to use custom masking in your Session Replays, you may accidentally expose sensitive customer data. Be sure to double-check what you choose to expose.

</Alert>

By default, our Session Replay SDK masks all text content, images, webviews and user input. This helps ensure that no sensitive data will be exposed. You can also manually choose which parts of your app's data you want to mask by using the different options listed below.

To disable the default masking behavior (not to be used on applications with sensitive data):

```kotlin
options.experimental.sessionReplay.maskAllText = false
options.experimental.sessionReplay.maskAllImages = false
```

![Session Replay unredacted](./img/session-replay.jpg) ![Session Replay redacted](./img/session-replay-redacted.jpg)

_Make sure your Sentry Android SDK version is at least 7.15.0._

## Mask by View Class

You can choose which type of view you want to mask or unmask by using the `addMaskViewClass` or `addUnmaskViewClass` options.

Let's say you have:
- A custom view that you want to mask

> - A `TextView` subclass (which normally would be masked) that you don't want to mask.

Reviewer: ```suggestion
- A `TextView` subclass (which normally would be masked) that you don't want to mask
```

You can set the options like this:

```kotlin
  options.experimental.sessionReplay.addMaskViewClass("com.example.MyCustomView")
  options.experimental.sessionReplay.addUnmaskViewClass("com.example.MyCustomTextView")
```

<Note>

> If you're using a code obfuscation tool (R8/ProGuard), adjust your proguard rules accordingly so your custom view class names are not minified.

Reviewer: ```suggestion
If you're using a code obfuscation tool (R8/ProGuard), adjust your proguard rules accordingly so your custom view class names don't get minified.
```

</Note>

### Class Hierarchy

> The masking behavior applies to classes and their subclasses. This means if you add a view via `addMaskViewClass` (for example, `TextView`, which is the default behavior), its respective subclasses (`RadioButton`, `CheckBox`, `EditText`, etc.) will also be masked. For example, you can do the following:

Reviewer: ```suggestion
The masking behavior applies to classes and their subclasses. This means if you add a view via `addMaskViewClass` (for example, `TextView`, which is the default behavior), its respective subclasses (`RadioButton`, `CheckBox`, `EditText`, and so on) will also be masked. For example, you can do the following:
```

```

## Mask by View Instance

You can also choose to mask or unmask a specific view instance by using tags like this:

```xml
  <View
    android:id="@+id/my_view"
    android:layout_width="wrap_content"
    android:layout_height="wrap_content"
    android:tag="sentry-mask|sentry-unmask"/>
```

```kotlin
  view.tag = "sentry-mask|sentry-unmask"
```

If your view already has a tag assigned, you can set the masking tag by a sentry-specific id:

```xml
  <View
    android:id="@+id/my_view"
    android:layout_width="wrap_content"
    android:layout_height="wrap_content">
    <tag android:id="@id/sentry_privacy" android:value="mask|unmask"/>
  </View>
```

```kotlin
  view.setTag(io.sentry.android.replay.R.id.sentry_privacy, "mask|unmask")
```

We also provide convenient extension functions for Kotlin:

```kotlin
  view.sentryReplayMask()
  // or
  view.sentryReplayUnmask()
```

## Jetpack Compose

We only support masking specific composables in Jetpack Compose. Since composables don't have a concept of classes (they are all composable functions), masking by view class is not supported.

> In this example, we want the "Hello" message to be captured in the replay (by default, all text composables are masked), but not the custom composable:

Reviewer: ```suggestion
In the below example, we want the "Hello" message to be captured in the replay, but not the custom composable. (By default, all text composables are masked.)
```

```kotlin
import io.sentry.android.replay.sentryReplayMask
import io.sentry.android.replay.sentryReplayUnmask

Column(
  verticalArrangement = Arrangement.Center,
  horizontalAlignment = Alignment.CenterHorizontally,
  modifier = Modifier.fillMaxSize()
) {
  MyCustomComposable(
    modifier = Modifier.fillMaxWidth().sentryReplayMask()
    ...
  )
  Text("Hello", modifier = Modifier.sentryReplayUnmask())
}
```

> Currently, we don't support masking anything within embedded android views (`AndroidView`), but you can still mask the entire view as follows:

Reviewer: ```suggestion
Currently, we don't support masking anything within embedded Android views (`AndroidView`), but you can still mask the entire view as follows:
```

```kotlin
view.sentryReplayMask()
// or
view.sentryReplayUnmask()
```

## Jetpack Compose

We only support masking specific composables in Jetpack Compose. Since composables are functions, not classes, masking by view class isn't possible. 

In the example below, we want the "Hello" message to be captured in the replay, but not the custom composable. (By default, all text composables are masked.)

```kotlin
import io.sentry.android.replay.sentryReplayMask
import io.sentry.android.replay.sentryReplayUnmask

Column(
  verticalArrangement = Arrangement.Center,
  horizontalAlignment = Alignment.CenterHorizontally,
  modifier = Modifier.fillMaxSize()
) {
  MyCustomComposable(
    modifier = Modifier.fillMaxWidth().sentryReplayMask()
    ...
  )
  Text("Hello", modifier = Modifier.sentryReplayUnmask())
}
```

Currently, we don't support masking anything within embedded Android views (`AndroidView`), but you can still mask the entire view as follows:

```kotlin
import io.sentry.android.replay.sentryReplayMask

AndroidView(
  modifier = Modifier.sentryReplayMask(),
  factory = { context -> ... }
)
```

## General Masking Rules

### View Groups

> - If a `ViewGroup` is marked as masked, **all its child views will also be masked**, even if some views would normally be ignored. This prioritizes safety and ensures no sensitive information is unintentionally exposed.

Author: ```suggestion
- If a `ViewGroup` is marked as masked, **all its child views will also be masked**, even if some views would normally not be masked. This prioritizes safety and ensures no sensitive information is unintentionally exposed.
```

---
title: Privacy
sidebar_order: 5501
notSupported:
description: "Learn how to mask parts of your app's data in Session Replay."
---

<Alert>

> 
> Before publising an App with Session Replay enabled, make sure to test it thoroughly to ensure that no sensitive data is exposed.
> 
> If you choose to use custom masking in your Session Replays, you may accidentally expose sensitive customer data. Be sure to double-check what you choose to expose.

Reviewer: ```suggestion

Using custom masking in your Session Replays may accidentally expose sensitive customer data. Before publishing an App with Session Replay enabled, make sure to test it thoroughly to ensure that no sensitive data is exposed.

```

</Alert>

> By default, our Session Replay SDK masks all text content, images, webviews and user input. This helps ensure that no sensitive data will be exposed. You can also manually choose which parts of your app's data you want to mask by using the different options listed below.
> 

Reviewer: ```suggestion
By default, our Session Replay SDK masks all text content, images, webviews, and user input. This helps ensure that no sensitive data is exposed. You can also manually choose which parts of your app's data you want to mask by using the different options listed below.

```

```kotlin
options.experimental.sessionReplay.addMaskViewClass("android.widget.TextView") // mask TextView and all its subclasses
options.experimental.sessionReplay.addUnmaskViewClass("android.widget.RadioButton") // but unmask RadioButton and all its subclasses
```

## Mask by View Instance

You can also choose to mask or unmask a specific view instance by using tags like this:

```xml
<View
  android:id="@+id/my_view"
  android:layout_width="wrap_content"
  android:layout_height="wrap_content"
  android:tag="sentry-mask|sentry-unmask"
/>
```

```kotlin
view.tag = "sentry-mask|sentry-unmask"
```

If your view already has a tag assigned, you can set the masking tag by a sentry-specific id:

```xml
<View
  android:id="@+id/my_view"
  android:layout_width="wrap_content"
  android:layout_height="wrap_content">

  <tag android:id="@id/sentry_privacy" android:value="mask|unmask"/>
</View>
```

```kotlin
  view.setTag(io.sentry.android.replay.R.id.sentry_privacy, "mask|unmask")
```

We also provide convenient extension functions for Kotlin:

```kotlin
view.sentryReplayMask()
// or
view.sentryReplayUnmask()
```

## Jetpack Compose

> 
> We only support masking specific composables in Jetpack Compose. Since composables don't have a concept of classes (they are all composable functions), masking by view class is not supported.
> 
> In the below example, we want the "Hello" message to be captured in the replay, but not the custom composable. (By default, all text composables are masked.)

Reviewer: ```suggestion

We only support masking specific composables in Jetpack Compose. Since composables are functions, not classes, masking by view class isn't possible. 

In the example below, we want the "Hello" message to be captured in the replay, but not the custom composable. (By default, all text composables are masked.)
```

```kotlin
import io.sentry.android.replay.sentryReplayMask
import io.sentry.android.replay.sentryReplayUnmask

Column(
  verticalArrangement = Arrangement.Center,
  horizontalAlignment = Alignment.CenterHorizontally,
  modifier = Modifier.fillMaxSize()
) {
  MyCustomComposable(
    modifier = Modifier.fillMaxWidth().sentryReplayMask()
    ...
  )
  Text("Hello", modifier = Modifier.sentryReplayUnmask())
}
```

Currently, we don't support masking anything within embedded Android views (`AndroidView`), but you can still mask the entire view as follows:

```kotlin
import io.sentry.android.replay.sentryReplayMask

AndroidView(
  modifier = Modifier.sentryReplayMask(),
  factory = { context -> ... }
)
```

> ## Masking Behavior

Author: @lizokm thanks for reviewing it, I added one more section down below, could you take a look please?

> ## Masking Behavior

Reviewer: ```suggestion
## General Masking Rules
```

> Here are some general rules we follow when applying masking rules:

Reviewer: ```suggestion

```

### View Groups

> - When a `ViewGroup` is marked as masked, **all its child views inherit this behavior and will be masked too**, even if some of those child views would typically be ignored. This approach prioritizes safety and ensures no sensitive information is exposed unintentionally.
> 
> - When a `ViewGroup` is marked as unmasked, **its child views do not automatically inherit its behavior**. Each child view must be explicitly marked as unmasked if you want to capture them in the replay.
> 

Reviewer: ```suggestion
- If a `ViewGroup` is marked as masked, **all its child views will also be masked**, even if some views would normally be ignored. This prioritizes safety and ensures no sensitive information is unintentionally exposed.

- If a `ViewGroup` is marked as unmasked, **its child views don't automatically inherit this behavior**. You'll need to explicitly mark each child view as unmasked if you want them to appear in the replay.

```

> ### Masking prioritization

Reviewer: ```suggestion
### Masking Priority 
```

> The following order determines how masking/unmasking rules are applied:
> 
>   1. Check if a view is marked as `unmasked` via a tag/extension function/modifier
>   2. Check if a view is marked as `masked` via a tag/extension function/modifier
>   3. Check if a view's class is marked as unmasked via `addUnmaskViewClass`
>   4. Check if a view's class is marked as masked via `addMaskViewClass`

Reviewer: ```suggestion
Masking and unmasking rules are applied in the following order:

  1. Check if a view is marked as `unmasked` via a tag/extension or function/modifier.
  2. Check if a view is marked as `masked` via a tag/extension or function/modifier.
  3. Check if a view's class is marked as unmasked via `addUnmaskViewClass`.
  4. Check if a view's class is marked as masked via `addMaskViewClass`.
```
